File: java/src/test/java/core/PhoneBookTest.java
package core;

import org.junit.jupiter.api.DisplayNameGeneration;
import org.junit.jupiter.api.DisplayNameGenerator.ReplaceUnderscores;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.stream.Collectors;

import static org.assertj.core.api.Assertions.assertThat;

@DisplayNameGeneration(ReplaceUnderscores.class)
class PhoneBookTest {

    private static final String MY_CONTACT_CSV = "my-contacts.csv";

    @Nested
    class slow {
        @Test
        void should_find_name_matching_phone_number_on_data() {
            // Arrange
            PhoneBook phoneBook = new PhoneBook(retrieveContacts());

            // Act
            String name = phoneBook.reverseLookup("0699999999");

            // Assert
            assertThat(name).isEqualTo("Jacques R");
        }
    }

    @Nested
    class fast {

        @Test
        void should_find_name_matching_phone_number_on_data() {
            // Arrange
            List<Contact> contacts = sampleContacts();
            PhoneBook phoneBook = new PhoneBook(contacts);

            // Act
            String name = phoneBook.reverseLookup("0612345678");

            // Assert
            assertThat(name).isEqualTo("Jacques");
        }

    }

    @Nested
    class not_self_validating {

        @Test
        void should_find_a_name_matching_phone_number_on_data() {
            // Arrange
            List<Contact> contacts = sampleContacts();
            PhoneBook phoneBook = new PhoneBook(contacts);

            // Act
            String name = phoneBook.reverseLookup("0612345678");

            // Assert
            System.out.println("nom du contact obtenu en cherchant avec 0612345678 est " + name);
        }

    }

    @Nested
    class other {

        @Test
        void should_find_all_contact_matching() {
            // Arrange
            PhoneBook phoneBook = new PhoneBook(retrieveContacts());

            // Act
            List<Contact> matchingContacts = phoneBook.findAllContact("Ja");

            // Assert
            assertThat(matchingContacts.size()).isEqualTo(266434);
        }

        @Test
        void should_add_new_contact() {
            // Arrange
            PhoneBook phoneBook = new PhoneBook(retrieveContacts());
            Contact newContact = new Contact("Jack Napier", "0777777777");

            // Act
            boolean contactWasAdded = phoneBook.addContact(newContact);
            if (contactWasAdded) {
                storeContacts(phoneBook.getContacts());
            }

            // Assert
            assertThat(contactWasAdded).isTrue();
        }

    }

    private List<Contact> retrieveContacts() {
        List<String> strings = new Persistence().retrieveFromPersistence(MY_CONTACT_CSV);
        return strings.stream()
                .map(line -> line.split(";"))
                .filter(tuple -> tuple.length == 2)
                .map(tuple -> new Contact(tuple[1], tuple[0]))
                .collect(Collectors.toList());
    }

    private void storeContacts(List<Contact> contacts) {
        new Persistence().storeToPersistence(contacts, PhoneBookTest.MY_CONTACT_CSV);
    }

    private List<Contact> sampleContacts() {
        Contact contact1 = new Contact("Jacques", "0612345678");
        Contact contact2 = new Contact("Michael", "0612345679");

        return List.of(contact1, contact2);
    }

}

File: js/test/PhoneBookTest.ts
var assert = require('assert');
var should = require('chai').should()
import PhoneBook from '../prod/PhoneBook'
import Contact from '../prod/Contact'
import Persistence from '../prod/Persistence';
let MY_CONTACT_CSV: string = "my-contacts.csv";

describe('PhoneBookTest', function() {


    describe('slow', function() {
        it('should find name matching phone number on data', function() {
            // Arrange
            let phoneBook: PhoneBook = new PhoneBook(retrieveContacts());

            // Act
            let name: string = phoneBook.reverseLookup("0699999999");

            // Assert
            name.should.equal("Jacques R");
        });
    });

    describe('fast', function() {
        it('should find name matching phone number on data', function() {
            // Arrange
            let contacts : Array<Contact> = sampleContacts();
            let phoneBook: PhoneBook = new PhoneBook(contacts);

            // Act
            let name: string = phoneBook.reverseLookup("0612345678");

            // Assert
            name.should.equal("Jacques");
        })
    })

    describe('not_self_validating', function() {
        it('should find name matching phone number on data', function() {
           // Arrange
           let contacts : Array<Contact> = sampleContacts();
           let phoneBook :PhoneBook = new PhoneBook(contacts);

           // Act
           let name :string = phoneBook.reverseLookup("0612345678");

           // Assert
           console.log("nom du contact obtenu en cherchant avec 0612345678 est " + name);
        })
    })

    describe('other', function() {
        it('should find all contact matching', function() {
            // Arrange
            let phoneBook : PhoneBook= new PhoneBook(retrieveContacts());

            // Act
            let matchingContacts : Array<Contact> = phoneBook.findAllContact("Ja");

            // Assert
            matchingContacts.size().should.equal(266434);
        }),

        it('should_add_new_contact', function() {
            // Arrange
            let phoneBook : PhoneBook= new PhoneBook(retrieveContacts());
            let newContact : Contact = new Contact("Jack Napier", "0777777777");

            // Act
            let contactWasAdded  : boolean = phoneBook.addContact(newContact);
            if (contactWasAdded) {
                storeContacts(phoneBook.getContacts());
            }

            // Assert
            contactWasAdded.should.equal(true);
         }),
         
    })
    function retrieveContacts(): Array<Contact>  {
        let strings: Array<String> = new Persistence().retrieveFromPersistence(MY_CONTACT_CSV);
        return strings.map(line => line.split(";"))
                .filter(tuple => tuple.length == 2)
                .map(tuple => new Contact(tuple[1], tuple[0]));
    }

    function storeContacts(contacts :Array<Contact> ) {
        new Persistence().storeToPersistence(contacts, MY_CONTACT_CSV);
    }

    function sampleContacts(): Array<Contact>  {
        let contact1:Contact = new Contact("Jacques", "0612345678");
        let contact2:Contact = new Contact("Michael", "0612345679");

        return Array.of(contact1, contact2);
    }
})